### PR TITLE
Playwright DDEV workflow

### PR DESCRIPTION
**Pull request by AlexP11223**
Adds a reusable workflow for running Playwright using the project's DDEV setup. Also optionally can start Ngrok. Runs successfully in [ddev-wordpress-plugin-template GHA](https://github.com/inpsyde/ddev-wordpress-plugin-template/actions/runs/4627008317/jobs/8184407364) See the [docs](https://github.com/inpsyde/reusable-workflows/blob/feature/ddev-playwright/docs/ddev-playwright.md) for more info.

The commands for setting up DDEV and running Playwright can be changed via parameters, because different projects do it differently. Once we have some kind of standard, we can add default values.

The parts about DDEV and Ngrok potentially can be extracted, but I think it must be actions, not reusable workflows, because you cannot run multiple reusable workflows inside a single environment.

### CODE REVIEW

**Comment by Chrico** [comment]
We're currently discussing this internally and also waiting for https://github.com/WordPress/gutenberg/pull/43998 to be published as package.

**Comment by tyrann0us** [comment]
FYI, `@wordpress/e2e-test-utils-playwright` is now available as a public package; see https://www.npmjs.com/package/@wordpress/e2e-test-utils-playwright.

**Comment by AlexP11223** [comment]
The package seems quite basic, maybe some useful things about managing users, posts, making API requests. https://github.com/WordPress/gutenberg/tree/a993bfd8e108561bc2ca37a13415847d87c2019f/packages/e2e-test-utils-playwright/src/request-utils

But anyway it's up to the tests whether to use something like this, it should not affect this workflow, it only setups DDEV, Ngrok, runs Playwright. It should be even possible to reuse for Cypress or other tools simply by override the Playwright install/run commands.

File: .github/workflows/ddev-playwright.yml [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+                core.setSecret(envVar.name);
+                core.setSecret(envVar.value);
+                exec.exec('ddev config --web-environment-add="' + envVar.name + '=' + envVar.value + '"');
+              });
+
+      - name: Add COMPOSER_AUTH into DDEV
+        env:
+          COMPOSER_AUTH_JSON: '${{ secrets.COMPOSER_AUTH_JSON }}'
+        run: echo '${{ env.COMPOSER_AUTH_JSON }}' > $GITHUB_WORKSPACE/auth.json
+        if: env.COMPOSER_AUTH_JSON
```
Comment (tyrann0us) [inline .github/workflows/ddev-playwright.yml:109]:
```suggestion
        if: env.COMPOSER_AUTH_JSON
        run: echo '${{ env.COMPOSER_AUTH_JSON }}' > $GITHUB_WORKSPACE/auth.json
```

File: docs/ddev-playwright.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+The "host" env variables (`ENV_VARS`) can be used in the tests (Playwright runs outside of DDEV), such as user credentials, card numbers.
+To keep things simple and avoid passing hundreds of variables,
+consider using some kind of `.env` file, and commit `.env.example` with usable defaults for parameters that do not have to be secret,
+then add `cp .env.example .env` in `PLAYWRIGHT_INSTALL_CMD` (`ddev pw-install-host` already includes it).
+
+The DDEV env variables (`DDEV_ENV_VARS`) can be used for debug flags of your project needed for the tests to run properly 
+(e.g. using placeholder product images or non-live mode).
+
+Also, it is possible to change some DDEV config values, such as the PHP version (`PHP_VERSION`).
```
Comment (tyrann0us) [inline docs/ddev-playwright.md:23]: Please check if it's still factually correct.
```suggestion
# Run Playwright tests using the project's DDEV setup

This reusable workflow:

1. Launches DDEV.
2. Runs the provided command such as `ddev orchestrate` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template) to set up the environment (install WP, plugins, ...).
3. If the Ngrok auth token is provided, it sets up Ngrok. Ngrok can be needed if some third-party service, e.g., webhooks, must access the website.
   1. Launches Ngrok, using `vendor/bin/ddev-share` from [inpsyde/ddev-tools](https://github.com/inpsyde/ddev-tools) by default.
   2. Saves the URL to the specified env variable, `BASEURL` by default.
4. Installs Playwright and its dependencies via the provided command, such as `yarn install && yarn playwright install --with-deps` or `ddev pw-install-host` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template).
5. Runs Playwright tests via the provided command, such as `yarn playwright test` or `ddev pw-host test` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template).

It is possible to add any env variables for the "host" and for DDEV.

The "host" env variables (`ENV_VARS`) can be used in the tests (Playwright runs outside of DDEV), such as user credentials or credit card numbers.
To keep things simple and avoid passing hundreds of variables,
consider using an `.env` file, and commit `.env.example` with usable defaults for parameters that do not have to be secret,
then add `cp .env.example .env` in `PLAYWRIGHT_INSTALL_CMD` (`ddev pw-install-host` already includes it).

Use the DDEV env variables (`DDEV_ENV_VARS`) for debug flags of your project needed for the tests to run properly 
(e.g., using placeholder product images or non-live mode).

Also, it is possible to change some DDEV config values, such as the PHP version (`PHP_VERSION`).
```

File: docs/ddev-playwright.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+| `PLAYWRIGHT_RUN_CMD`     | `""`                      | The command for running Playwright tests, such as `yarn playwright test` or `ddev pw-host test` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template).                                             |
+| `NGROK_START_CMD`        | `"vendor/bin/ddev-share"` | The command for starting Ngrok, such as `ddev-share` from [inpsyde/ddev-tools](https://github.com/inpsyde/ddev-tools).                                                                                                                                |
+| `BASEURL_ENV_NAME`       | `"BASEURL"`               | The name of the env variable with the base URL for Playwright. Used for overwriting it with the URL from Ngrok.                                                                                                                                       |
```
Comment (tyrann0us) [inline docs/ddev-playwright.md:64]: Please check if the Markdown table syntax is still correct after applying this change.
```suggestion
| `PHP_VERSION`            | `""`                      | PHP version which will override the version set in the DDEV config                                                                                                                                                                                   |
| `NODE_VERSION`           | `""`                      | Node version which will override the version set in the DDEV config                                                                                                                                                                                  |
| `DDEV_ORCHESTRATE_CMD`   | `""`                      | The command for setting up the DDEV website, such as `ddev orchestrate` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template)                                                                     |
| `PLAYWRIGHT_INSTALL_CMD` | `""`                      | The command for installing Playwright and its deps, such as `yarn install && yarn playwright install --with-deps` or `ddev pw-install-host` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template) |
| `PLAYWRIGHT_RUN_CMD`     | `""`                      | The command for running Playwright tests, such as `yarn playwright test` or `ddev pw-host test` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template)                                              |
| `NGROK_START_CMD`        | `"vendor/bin/ddev-share"` | The command for starting Ngrok, such as `ddev-share` from [inpsyde/ddev-tools](https://github.com/inpsyde/ddev-tools)                                                                                                                                 |
| `BASEURL_ENV_NAME`       | `"BASEURL"`               | The name of the env variable with the base URL for Playwright. Used for overwriting it with the URL from Ngrok                                                                                                                                       |
```

File: docs/ddev-playwright.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+## Secrets
+
+| Name                 | Description                                                                               |
+|----------------------|-------------------------------------------------------------------------------------------|
+| `NPM_REGISTRY_TOKEN` | Authentication for the private npm registry.                                              |
+| `COMPOSER_AUTH_JSON` | Authentication for privately hosted packages and repositories as a JSON formatted object. |
+| `ENV_VARS`           | Additional environment variables as a JSON formatted object.                              |
+| `DDEV_ENV_VARS`      | Additional environment variables for DDEV as a JSON formatted object.                     |
+| `NGROK_AUTH_TOKEN`   | The auth token for Ngrok. If not provided, skips Ngrok.                                   |
```
Comment (tyrann0us) [inline docs/ddev-playwright.md:74]:
```suggestion
| `COMPOSER_AUTH_JSON` | Authentication for privately hosted packages and repositories as a JSON formatted object |
| `NPM_REGISTRY_TOKEN` | Authentication for the private npm registry                                              |
| `NGROK_AUTH_TOKEN`   | Ngrok auth token; skips the installation of Ngrok if not provided                        |
| `ENV_VARS`           | Additional environment variables as a JSON formatted object                              |
| `DDEV_ENV_VARS`      | Additional environment variables for DDEV as a JSON formatted object                     |
```

**Comment by Chrico** [comment]
In our meeting today (16.01.2024) we discussed about the possiblities of running this reusable-workflow in a Matrix. So having different PHP and WordPress version. Is this possible at Github Action level or do we need to have additional configuration in DDEV (like different containers for different PHP/WordPress-versions)?

**Comment by AlexP11223** [comment]
If using DDEV, then it needs to be configured to use the desired WP/WC/PHP version. Like here https://github.com/woocommerce/woocommerce-paypal-payments/blob/trunk/.github/workflows/e2e.yml

File: .github/workflows/ddev-playwright.yml [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+      DDEV_ORCHESTRATE_CMD:
+        description: The command for setting up the DDEV website, such as 'ddev orchestrate' from inpsyde/ddev-wordpress-plugin-template.
+        required: false
+        type: string
+      PLAYWRIGHT_INSTALL_CMD:
+        description: The command for installing Playwright and its deps, such as `yarn install && yarn playwright install --with-deps` or 'ddev pw-install-host' from inpsyde/ddev-wordpress-plugin-template.
+        required: true
+        type: string
+      PLAYWRIGHT_RUN_CMD:
+        description: The command for running Playwright tests, such as `yarn playwright test` or 'ddev pw-host test' from inpsyde/ddev-wordpress-plugin-template.
```
Comment (tyrann0us) [inline .github/workflows/ddev-playwright.yml:23]:
```suggestion
        description: The command for setting up the DDEV website, such as `ddev orchestrate` from inpsyde/ddev-wordpress-plugin-template.
        required: false
        type: string
      PLAYWRIGHT_INSTALL_CMD:
        description: The command for installing Playwright and its dependencies, such as `yarn install && yarn playwright install --with-deps` or `ddev pw-install-host` from inpsyde/ddev-wordpress-plugin-template.
        required: true
        type: string
      PLAYWRIGHT_RUN_CMD:
        description: The command for running Playwright tests, such as `yarn playwright test` or `ddev pw-host test` from inpsyde/ddev-wordpress-plugin-template.
```

File: .github/workflows/ddev-playwright.yml [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        required: true
+        type: string
+      PLAYWRIGHT_DIR:
+        description: The path to the Playwright project.
+        required: false
+        default: 'tests/Playwright'
+        type: string
+      NGROK_START_CMD:
+        default: 'vendor/bin/ddev-share'
+        description: The command for starting Ngrok, such as ddev-share from inpsyde/ddev-tools.
```
Comment (tyrann0us) [inline .github/workflows/ddev-playwright.yml:33]:
```suggestion
        description: The command for starting Ngrok, such as `ddev-share` from inpsyde/ddev-tools.
```

File: docs/ddev-playwright.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+| `PLAYWRIGHT_RUN_CMD`     | `""`                      | The command for running Playwright tests, such as `yarn playwright test` or `ddev pw-host test` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template)                                              |
+| `NGROK_START_CMD`        | `"vendor/bin/ddev-share"` | The command for starting Ngrok, such as `ddev-share` from [inpsyde/ddev-tools](https://github.com/inpsyde/ddev-tools)                                                                                                                                 |
+| `BASEURL_ENV_NAME`       | `"BASEURL"`               | The name of the env variable with the base URL for Playwright. Used for overwriting it with the URL from Ngrok                                                                                                                                       |
```
Comment (tyrann0us) [inline docs/ddev-playwright.md:65]:
```suggestion
| `BASEURL_ENV_NAME`       | `"BASEURL"`               | The name of the env variable with the base URL for Playwright, used for overwriting it with the URL from Ngrok                                                                                                                                       |
```

File: docs/ddev-playwright.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+## Secrets
+
+| Name                 | Description                                                                               |
+|----------------------|-------------------------------------------------------------------------------------------|
+| `COMPOSER_AUTH_JSON` | Authentication for privately hosted packages and repositories as a JSON formatted object |
+| `NPM_REGISTRY_TOKEN` | Authentication for the private npm registry                                              |
+| `NGROK_AUTH_TOKEN`   | Ngrok auth token; skips the installation of Ngrok if not provided                        |
+| `ENV_VARS`           | Additional environment variables as a JSON formatted object                              |
+| `DDEV_ENV_VARS`      | Additional environment variables for DDEV as a JSON formatted object                     |
+| `SSH_KEY`            | The private SSH key allowing access remote destinations.                                  |
```
Comment (tyrann0us) [inline docs/ddev-playwright.md:76]:
```suggestion
| `SSH_KEY`            | Private SSH key to be used to reach remote destinations                                   |
```

File: docs/ddev-playwright.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+| `PLAYWRIGHT_INSTALL_CMD` | `""`                      | The command for installing Playwright and its deps, such as `yarn install && yarn playwright install --with-deps` or `ddev pw-install-host` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template) |
+| `PLAYWRIGHT_RUN_CMD`     | `""`                      | The command for running Playwright tests, such as `yarn playwright test` or `ddev pw-host test` from [inpsyde/ddev-wordpress-plugin-template](https://github.com/inpsyde/ddev-wordpress-plugin-template)                                              |
+| `NGROK_START_CMD`        | `"vendor/bin/ddev-share"` | The command for starting Ngrok, such as `ddev-share` from [inpsyde/ddev-tools](https://github.com/inpsyde/ddev-tools)                                                                                                                                 |
```
Comment (tyrann0us) [inline docs/ddev-playwright.md:64]: Please add `PLAYWRIGHT_DIR`.

File: .github/workflows/ddev-playwright.yml [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        description: Additional environment variables as a JSON formatted object.
+        required: false
+      DDEV_ENV_VARS:
+        description: Additional environment variables for DDEV as a JSON formatted object.
+        required: false
+      SSH_KEY:
+        description: Private SSH key to be used to reach remote destinations.
+        required: false
+      SSH_KNOWN_HOSTS:
+        description: SSH hosts to be set in the known_hosts file.
```
Comment (tyrann0us) [inline .github/workflows/ddev-playwright.yml:61]:
```suggestion
        description: SSH hosts to be set in the `known_hosts` file.
```

File: docs/ddev-playwright.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+| Name                 | Description                                                                               |
+|----------------------|-------------------------------------------------------------------------------------------|
+| `COMPOSER_AUTH_JSON` | Authentication for privately hosted packages and repositories as a JSON formatted object |
+| `NPM_REGISTRY_TOKEN` | Authentication for the private npm registry                                              |
+| `NGROK_AUTH_TOKEN`   | Ngrok auth token; skips the installation of Ngrok if not provided                        |
+| `ENV_VARS`           | Additional environment variables as a JSON formatted object                              |
+| `DDEV_ENV_VARS`      | Additional environment variables for DDEV as a JSON formatted object                     |
+| `SSH_KEY`            | The private SSH key allowing access remote destinations.                                  |
+| `SSH_KNOWN_HOSTS`    | SSH hosts to be set in the `known_hosts` file.                                            |
```
Comment (tyrann0us) [inline docs/ddev-playwright.md:77]:
```suggestion
| `SSH_KNOWN_HOSTS`    | SSH hosts to be set in the `known_hosts` file                                             |
```

**Review by tyrann0us — COMMENTED** [review]
I left another (hopefully last) round of comments regarding the input descriptions. Once you are done, please reformat the markdown table. Thanks!

Apart from that, I'm unsure regarding the support for backticks (`` ` ``) in YAML files; could you please test if they are allowed and valid?

**Comment by AlexP11223** [comment]
Backticks should be ok if not at the start of a string without quotes. Not sure where GH displays these descriptions and if they are formatted, but even if displayed as text it is nice to separate commands.

**Review by tyrann0us — APPROVED** [review]
Thanks for incorporating all the changes. LGTM. Great work!

**Pull request merged**